### ISSUE TITLE
Concurrent resolving of open generic registrations with HierarchicalLifetimeManager

### ISSUE THREAD

**Issue by Burnaster** [unitycontainer/unity#56]
Concurrent resolving of open generic registrations
### Situation

In our application, we are using the generic repository pattern. This means, we have a generic interface, which is implemented for all of our entity classes (i.e. one repository for each type). We used the standard open generic registration with a hierarchical lifetime manager:

```
// Register Generic interface with Generic implementation --> Does Pronot correctly resolve.
parentContainer.RegisterType(typeof(IRepository<>), typeof(GenericRepository<>), 
new HierarchicalLifetimeManager());

```

In our application, we have various ASP.NET Web API methods. The front-end, written in HTML5 and AngularJS, is capable of firing several requests in parallel.
### The error

We got occasional InvalidOperationExceptions “The DbContext has been disposed”. This happened when we tried to perform any WebAPI operation that involved one of our repositories.
When investigating, we found several conditions that have to be satisfied for having a chance that the error actually occurs:
-   Multiple CPUs: When we had only one CPU on our virtual server, the error never occurred. It only started when we got 2 and more CPUs.
-   Parallel requests: When the front-end only sends one request at a time, or waits long enough (100ms) between two requests, the error did not occur.
-   Only the first time: The error occurred only when we just started the IIS worker process, but never when it was already up and running.
### Workaround

Explicit registrations of generic types avoids the above mentioned behaviour.
`
parentContainer.RegisterType<IRepository<SomeEntity>, GenericRepository<SomeEntity>>(new HierarchicalLifetimeManager());
`
### Reproduction

I attached a c# console Project that can be used to reproduce the described issue.
[GenericDependencyResolve.zip](https://github.com/unitycontainer/unity/files/286599/GenericDependencyResolve.zip)

**Comment by Kylar** [comment]
It seems like #37 is the root problem here.

**Comment by ENikS** [comment]
duplicate of #37
